loginserver.js

```diff
@@ -17,20 +17,45 @@ const url = require('url');
 
 const FS = require('./fs');
 
-let TimeoutError = function (message) {
-	Error.captureStackTrace(this, TimeoutError);
-	this.name = "TimeoutError";
-	this.message = message || "";
-};
-TimeoutError.prototype = Object.create(Error.prototype);
-TimeoutError.prototype.constructor = TimeoutError;
-TimeoutError.prototype.toString = function () {
-	if (!this.message) return this.name;
-	return this.name + ": " + this.message;
-};
+/**
+ * A custom error thrown when a problem specific to the login server occurs,
+ * rather than the number of generic errors Node uses. Error's internal
+ * behaviour makes subclassing it require special treatment so it can act like
+ * an Error properly.
+ */
+class TimeoutError extends Error {
+	/**
+	 * Assigning this in the constructor would make the error print extra
+	 * properties unrelated to its stack.
+	 *
+	 * @return {string}
+	 */
+	get name() {
+		return 'TimeoutError';
+	}
+
+	/**
+	 * Stops instances from returning 'Error' when stringified.
+	 *
+	 * @return {string}
+	 */
+	get [Symbol.toStringTag]() {
+		return this.name;
+	}
+
+	/**
+	 * Stops instances from not being considered to be their actual type.
+	 *
+	 * @return {{new(message: [string]): TimeoutError}}
+	 */
+	static get [Symbol.species]() {
```

Comment:
This is overridden because `LoginServer#requestEnd` actually does has a condition if an instance of `TimeoutError` is passed to it. Not doing this would prevent the request log from recording timeouts